feat: pre-request estimated cost tracking

voodoohop opened the pull request:
- Add `pending_estimate` event status for in-flight requests
- Add `estimatedCost` column to event table
- Insert estimate event BEFORE request, UPDATE after completion
- Include estimated costs in balance calculation (`getRecentSpend`)
- Filter `pending_estimate` from Polar/Tinybird delivery
- Self-healing: stale estimates expire after 10min window

Addresses #6083

voodoohop committed:
feat: Show real model costs from Tinybird in pricing table

- Add /api/model-stats endpoint with KV caching (1h TTL)
- Fetch actual avg_cost_usd from Tinybird public_model_stats pipe
- Update pricing calculations to use real stats when available
- Fix rounding: show 1 decimal for numbers under 5
- Fix video column header: 'videos' instead of 'seconds'
- Update Tinybird query to use toStartOfDay for stable daily results

voodoohop committed:
feat: pre-request estimated cost tracking to prevent parallel request abuse

- Add pending_estimate event status for in-flight requests
- Add estimatedCost column to event table (migration 0008)
- Insert pending_estimate event BEFORE request, UPDATE after completion
- Include estimatedCost in getRecentSpend() balance calculation
- Filter pending_estimate events from Polar/Tinybird delivery
- Add model-stats utility for fetching average costs from Tinybird

Addresses #6083

voodoohop committed:
test: add test for pending_estimate event filtering

voodoohop committed:
fix: include estimatedCost in dashboard pending spend display

voodoohop committed:
simplify: build Tinybird URL inline instead of utility function

voodoohop committed:
refactor: simplify model-stats URL to use hardcoded public Tinybird endpoint

- Remove dynamic URL building from env vars
- Use public Tinybird pipe URL directly (read-only token)
- Simplify getModelStats function signature

eulervoid commented on enter.pollinations.ai/src/middleware/track.ts:
```diff
@@ -120,6 +122,67 @@ export const track = (eventType: EventType) =>
+                    resolvedModelRequested:
+                        requestTracking.resolvedModelRequested,
+                    modelProviderUsed: requestTracking.modelProvider,
+                    isBilledUsage: true, // Assume billable, will be updated
```
I think the default should be false here.

voodoohop committed:
refactor: simplify model-stats and add public token comment

- Add comment explaining Tinybird token is public/read-only
- Simplify getEstimatedCost to return 0 when no stats (removes default estimate)

eulervoid reviewed (commented):
I like the general idea, but i'm not sure about updating the events, in my head they were immutable. But i guess we already update them when they change status and its fine, but i'd want fields like `totalCost` to be undefined while the event is incomplete to make it really clear that this is whats going on, and to make more modular functions and types. Maybe i could have a go at this and try to sort things a bit?

voodoohop committed:
fix: pending_estimate events use isBilledUsage=false and undefined costs

- Set isBilledUsage to false for pending_estimate events (unknown until complete)
- Remove totalCost/totalPrice from pending events (leave undefined)
- Makes incomplete event state explicit

voodoohop committed:
refactor: rename estimatedCost to estimatedPrice

- Rename DB column estimated_cost → estimated_price (migration 0009)
- Rename schema field estimatedCost → estimatedPrice
- Rename Tinybird field avg_cost_usd → avg_price_usd
- Rename function getEstimatedCost → getEstimatedPrice
- Update all SQL queries and tests

This clarifies that the field stores the estimated price (what we charge users),
not cost (what we pay providers).

voodoohop committed:
Merge main into feat/pre-request-cost-tracking

- Resolve conflicts in routes, middleware, and drizzle migrations
- Update getPendingSpend() to include estimatedPrice for pending_estimate events
- Add migration for estimated_price column
- Clean up unused imports

voodoohop committed:
snapshots
